packages/core/content-type-builder/server/src/services/components.ts

```diff
@@ -86,13 +86,38 @@ export const editComponent = async (
   return updatedComponent;
 };
 
-export const deleteComponent = async (uid: Internal.UID.Component) => {
-  const builder = createBuilder();
+export const deleteComponent = async (deleteUid: Internal.UID.Component) => {
+  const deletedComponent = await strapi.db.transaction(async ({ trx }) => {
+    const builder = createBuilder();
 
-  const deletedComponent = builder.deleteComponent(uid);
+    const models = [...builder.contentTypes.entries(), ...builder.components.entries()];
 
-  await builder.writeFiles();
+    // Find models that include an attribute with this component as a target and delete references to it from the db
+    for (const [modelUid] of models) {
+      const metadata = strapi.db.metadata.get(modelUid);
+
+      const matchingAttributes = Object.values(metadata.attributes).filter(
+        (attr: any) => attr.target === deleteUid && attr.joinTable?.name
+      );
+
+      // Delete entries in each join table associated with matching attributes
+      for (const attr of matchingAttributes) {
+        if (!('joinTable' in attr && attr.joinTable && attr.joinTable.name)) {
+          continue;
+        }
+
+        await trx(attr.joinTable.name).where('component_type', deleteUid).delete();
```

Comment:
Any opinions on if this should be allowed to throw, or if this should be treated as a non-critical part of the delete and just warn? I already have a stashed change that just logs a warning that the data couldn't be deleted if there is a problem, which would make this change a lot safer in the case that we've missed any edge cases where we target the wrong table for example.

Reply:
I think this will rarely be usefull. the CTB is used locally so most likely not connected to a production database and as such will not change anything when the deployment happen. Considering this I would reco a warning

Reply:
True, this will just be preventing immediate errors. The migration or repair will be needed to fix it in production.